### ISSUE TITLE
feature: including external docker-compose.yml

### ISSUE THREAD

**Issue by dnephin** [docker/compose#318]
### The Problem

I'd like to use fig for acceptance testing of services

In the case of a single service with a single level of dependencies this is easy (ex: a webapp with a database). As soon as the dependency tree grows to depth > 1, it gets more complicated (ex: service-a requires service-b which requires a database). Currently I'd have to specify the service dependency tree in each `fig.yml`.  This is not ideal because as the tree grows, we end up with a lot of duplication, and having to update a few different `fig.yml`'s in different projects when a service adds/changes a dependency is not great.
### Proposed Solution

Support an `include` section in the `fig.yml` which contains url/paths to other `fig.yml` files. These files would be included in a `Project` so that the top-level `fig.yaml` can refer to them using `<proejct>_<service>`.
#### Example config

``` yaml
include:
    # link using http
    servicea:
        http: http://localhost:8080/projects/servicea/fig.yml
    # local paths
    serviceb:
        path: ../projects/serviceb/fig.yml

webapp:
    ...
    links:
        - servicea_webapp
        - serviceb_webapp
```

**Comment by sdake** [comment]
@dnephin Reviewed this patch looks pretty solid.  It needs a rebase, which will probably be some work.  There are a couple comments in the review.  I have also asked a good friend who is a python wizard to have a look over it @zaneb.

**Comment by erichonkanen** [comment]
it seems like this proposal may be what I was thinking in https://github.com/docker/compose/issues/1647

**Comment by dnephin** [comment]
I agree, I think they're trying to solve the same problem

**Comment by dnephin** [comment]
I've implemented this as an external pre-processing tool at https://github.com/dnephin/compose-addons#dcao-include

If anyone is interested, please check out the docs. Any feedback can be contributed by opening an issue on that repo (and would be appreciated).

**Comment by yoanisgil** [comment]
@dnephin I'd like to give a try to your branch but I cannot seem to find it here:

https://github.com/dnephin/compose/branches

unfortunately I cannot switch to a different tool to orchastrate my containers so If you let me know how can I use your modifs I can give it a try and get you some feedback.

**Comment by dnephin** [comment]
The branch is so old now I don't think it will merge.

I would suggest giving  https://github.com/dnephin/compose-addons#dcao-include a try. It's a pre-processing step, but you still use `docker-compose` for the orchestration.  I've used `compose-addons` for a few repos and it's worked for me.

If the issue is a lack of a binary install, please do open an issue on the github for that project and I can help you out.

**Comment by yoanisgil** [comment]
Ok. Let me see if I can give it a try ;)

**Comment by DCameronMauch** [comment]
Did anything ever happen with this?  If not, will something happen?  We are having to do some ugly things with multiple "-f" and relative path stuff.

**Comment by dnephin** [comment]
I build some things external to Compose in  https://github.com/dnephin/compose-addons that act as pre/post processors for a Compose file. They probably need to be updated for the v2 format.

**Comment by proudcommerce** [comment]
+1

**Comment by tjamet** [comment]
@dnephin I like the idea of dcao include, do you think there is any chance to get it a part of compose itself?

**Comment by dnephin** [comment]
We have not made any progress integrating the ideas into Compose

**Comment by tjamet** [comment]
Is there any way we can help doing so? or shall we use the add-ons?

**Comment by nharward-hd** [comment]
Simplified example of a situation I've run into at multiple companies that I wish docker-compose could support and could with includes:

```
my_top_level_project_dir/docker-compose.yml
   db/docker-compose.yml
   middle-tier/docker-compose.yml
   front-end/docker-compose.yml
   another-front-end/docker-compose.yml
```

Each team/developer responsible for an area should be able to change to their appropriate subdir and type `docker-compose up` and be done with it. The middle-tier would include the db layer, and each front end would include the middle tier definitions. The top-level directory would start everything, by including each of the front-ends. We have a more complex setup with more components, but hopefully the idea gets across.

In this way each team can work [far] more efficiently by only using/defining the components they actually need. By allowing linking/depends_on for includes you can have a DRY system where each component can be effectively managed/versioned without the headache of managing multiple files, or multiple `-f` options on the command line.

**Comment by sgreenslade** [comment]
+1 for Include

**Comment by ludekvodicka** [comment]
Any news about this?

**Comment by jvanbruegge** [comment]
I really like the way the syntax works with `dcao-include`. Sadly it does not work with `V3`.

**Comment by jahbini** [comment]
Has anyone suggested just using the c-preprocessor parse a yml file just like a c.header file?  It would allow simple (left-justified) includes.  The ability to use it's macro capabilities might be useful too.

**Comment by pwFoo** [comment]
Also looking for a solution to use external docker-compose.yml by http/s url.

**Comment by rakshazi** [comment]
Hello!
Any updates on this?

**Comment by dosmanak** [comment]
This feature would really help us since we use a lot `docker-compose -f docker-compose-common.yml -f docker-compose-local.yml` and it sucks to write it down all the time.

**Comment by ludekvodicka** [comment]
As @dosmanak mentioned. This would be very appreciated feature.

**Comment by Papipo** [comment]
This is easily fixed by creating single-line shell scripts.

**Comment by ludekvodicka** [comment]
@Papipo: Everything can be fixed by shell scripts. 

We have Windows and Linux developer machines so we have to maintain two different scripts for each project. Instead of including one simple line in the docker-compose file.

**Comment by Papipo** [comment]
@ludekvodicka that's a good point :)

**Comment by jvanbruegge** [comment]
I also want this for dependency reasons. I want to be able to have a service depend on a recursive service compound.
I have a reverse proxy that needs to be started after a few others, but not necessarily at last. All the services it depends on are defined in docker-compose files in subdirectories. It would be cool if I could do it like this:
```
services:
     reverse-proxy:
         #more fields here
         depends_on:
             - logger
             - ./my-recusive-service:service1

    logger:
        #more here
```
So when `depends_on` gets a path, it waits for `service1` defined in `./my-recursive-service/docker-compose.yaml`

**Comment by ET-CS** [comment]
@jvanbruegge Be aware that `depends_on` doe's not wait for `mysql` to be "ready" thus should not be used in order to control startup order.

[https://docs.docker.com/compose/compose-file/#depends_on](https://docs.docker.com/compose/compose-file/#depends_on)

[https://docs.docker.com/compose/startup-order/](https://docs.docker.com/compose/startup-order/)

**Comment by sarink** [comment]
Having to type out `docker-compose -f docker-compose-common.yml -f docker-compose-local.yml <some command>` 500 times a day is just the worst.

Why in the world is this _still_ not implemented? (it's fundamentally different from yaml merges/anchors, so those are not a substitute)

**Comment by luishdez** [comment]
I thinks this issue can be close @dnephin  Since it's developed as an Add-on https://github.com/dnephin/compose-addons#dcao-include and won't be included in the main project.

**Comment by carlosonunez** [comment]
I disagree. Many templating languages support including sub-templates to discourage DRY and encourage better organization. I think Compose should as well.

**Comment by luishdez** [comment]
@carlosonunez I would love to see this include but after 4 years looks that it's not gonna happen anytime soon. 🤷‍♂️ And this issue has been followed in https://github.com/docker/compose/pull/758 

I mean just to keep issues clean and concise.

**Comment by pirate** [comment]
bump

**Comment by sarink** [comment]
I recently discovered the `COMPOSE_FILE` env var in combination with the `env_file: .env` command. What I do now, is have a main `docker-compose.yml`, and then extensions to it like `docker-compose.local.yml`, `docker-compose.prod.yml`, etc. Then, inside the `.env` file (which will be different for every environment), I write something like
```
COMPOSE_FILE=docker-compose.yml:docker-compose.local.yml
```
Now, all you need to do is have parallel env files, which you probably need anyway

**Comment by dosmanak** [comment]
That seems like it exactly implements the need for include docker compose
files. Thank you.
The link to docs: https://docs.docker.com/compose/reference/envvars/


Dne so 12. 10. 2019 16:00 uživatel Kabir Sarin <notifications@github.com>
napsal:

**Comment by reitzig** [comment]
Including YAML bits in general would be useful. For instance, say we have a port mapping

```yaml
ports:
    - "8080:80"
```

and want to turn it off in a CI build. There doesn't seem to be a way to do it. If we could write something like (inspired by Ansible playbooks)

```yaml
ports: {{ my_ports.yml }}
```

where `my_ports.yml` contained a YAML array, it would be possible.

**Comment by sarink** [comment]
@reitzig it's worth mentioning you could specify these in a `.env` file, which is an o-k workaround ime.

**Comment by reitzig** [comment]
@sarink And what exactly do I put into `.env` so that there's no error but no port exposed?

**Comment by tmstff** [comment]
I'd still love to have this feature!

**Comment by andyl** [comment]
This feature sure would be helpful.  How many years now?

**Comment by reitzig** [comment]
I find that using multiple, "layered" docker-compose.yml 80/20-solves this for me. It's not ideal, as I end up writing wrapper scripts around docker-compose (a wrapper), but well.

**Comment by henri9813** [comment]
Hello,

How an i include a compose file via HTTP ? 

docker-compose up -f http://test.fr/docker-compose.yml ?

That's a needed feature ! We can actually do docker-compose -f mycompose.yaml. 

Is it possible to add at least support for HTTP file protocol ?

**Comment by keemyb** [comment]
> @sarink And what exactly do I put into `.env` so that there's no error but no port exposed?

I haven't tried it, but you could probably do the following:

```
# docker-compose.yml
ports:
  - "${MY_PORT}"
```
```
# dev.env
MY_PORT=4000:4000
```
```
# prod.env
MY_PORT=4000
```

It's not great, but the docs https://docs.docker.com/compose/compose-file/#ports don't indicate there is a special value that maps the host port randomly. It only works in the absence of a host port by the looks of it.

**Comment by keemyb** [comment]
> Hello,
> 
> How can i include a compose file via HTTP ?
> 
> docker-compose up -f http://test.fr/docker-compose.yml ?
> 
> That's a needed feature ! We can actually do docker-compose -f mycompose.yaml.
> 
> Is it possible to add at least support for HTTP file protocol ?

Why not
```
curl http://test.fr/docker-compose.yml -O && docker-compose up -f ./docker-compose.yml 
```

**Comment by pirate** [comment]
bump

**Comment by henri9813** [comment]
@keemyb , yes, that would work, but pretty dirty :/ in my opinion.

**Comment by jamshid** [comment]
This issue has a lot of different ideas but fwiw you can read the compose file from stdin with `-f -`. Reading directly from http would be nice, especially when there are multiple compose files, but at least this avoids a temporary file.
```curl -fsS http://test.fr/docker-compose.yml - | docker-compose up -f -```

**Comment by pirate** [comment]
Boop

**Comment by jgentil** [comment]
This isn't stale.  C'mon, this is dead simple.  We need a sensible way to combine and chain yml files together.  Even a YAML directive like !include would be great.

**Comment by bsdis** [comment]
This is something that would be very good to have

+1

**Comment by thaJeztah** [comment]
I think for local files, this would (to some extend) be possible already through the `extends` option; https://docs.docker.com/compose/compose-file/#extends

```bash
tree
.
├── backend
│   └── docker-compose.yaml
├── docker-compose.yaml
└── frontend
    └── docker-compose.yaml
```

frontend/docker-compose.yaml:

```yaml
services:
  web:
    image: nginx:alpine
```

backend/docker-compose.yaml:

```yaml
services:
  db:
    image: mysql:latest
```

docker-compose.yaml:

```yaml
services:
  web:
    extends:
      file: frontend/docker-compose.yaml
      service: web
  db:
    extends:
      file: backend/docker-compose.yaml
      service: db
```

```bash
docker compose up
[+] Running 3/3
 ⠿ Network compose_extend_default  Created                                                                                                                                                                                                                0.0s
 ⠿ Container compose_extend-web-1  Created                                                                                                                                                                                                                0.0s
 ⠿ Container compose_extend-db-1   Created                                                                                                                                                                                                                0.0s
Attaching to compose_extend-db-1, compose_extend-web-1
compose_extend-web-1  | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
compose_extend-web-1  | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
compose_extend-web-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
compose_extend-web-1  | 10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
compose_extend-web-1  | 10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
compose_extend-web-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
compose_extend-web-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
compose_extend-web-1  | /docker-entrypoint.sh: Configuration complete; ready for start up
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: using the "epoll" event method
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: nginx/1.23.3
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: built by gcc 12.2.1 20220924 (Alpine 12.2.1_git20220924-r4)
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: OS: Linux 5.15.49-linuxkit
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker processes
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker process 30
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker process 31
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker process 32
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker process 33
compose_extend-web-1  | 2023/01/27 14:54:27 [notice] 1#1: start worker process 34
compose_extend-db-1   | 2023-01-27 14:54:27+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.32-1.el8 started.
compose_extend-db-1   | 2023-01-27 14:54:27+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
compose_extend-db-1   | 2023-01-27 14:54:27+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.32-1.el8 started.
compose_extend-db-1   | 2023-01-27 14:54:27+00:00 [ERROR] [Entrypoint]: Database is uninitialized and password option is not specified
compose_extend-db-1   |     You need to specify one of the following as an environment variable:
compose_extend-db-1   |     - MYSQL_ROOT_PASSWORD
compose_extend-db-1   |     - MYSQL_ALLOW_EMPTY_PASSWORD
compose_extend-db-1   |     - MYSQL_RANDOM_ROOT_PASSWORD
compose_extend-db-1 exited with code 1
```

**Comment by theoilie** [comment]
this still would be a great feature to have 😄

**Comment by ndeloof** [comment]
Working on [definition](https://github.com/compose-spec/compose-spec/pull/363) for this in compose file format, we (the Docker Team) can't establish a consensus on the right term to use for this feature.

Tell us the one you prefer:

/polls "require" "import" "include"

**Comment by pirate** [comment]
Very exciting! I hope someday this could help us towards having a Docker Hub equivalent for sharing Compose files. Import features would be very powerful if you could import / depend on a compose project from the "Hub" just like you can base Dockerfiles on images from Docker Hub.

**Comment by ndeloof** [comment]
@pirate you'd be interested by https://github.com/compose-spec/compose-spec/issues/288 then

**Comment by mbrevda** [comment]
@ndeloof to clarify: does this issue being closed mean that `include` is available in the latest `docker compose`?

**Comment by ndeloof** [comment]
@mbrevda it will be "included" (🤪)  in next release, which should be v2.20.0

**Comment by mbrevda** [comment]
Looks like this didn't make the cut? Or did it just not make the release notes?
